p2p/transport/pipe/pipestream.go

```diff
@@ -0,0 +1,77 @@
+package pipetransport
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	streammux "github.com/libp2p/go-stream-muxer"
+)
+
+var ErrHalfClosed = fmt.Errorf("tried to write to stream that was half closed")
+var ErrAlreadyHalfClosed = fmt.Errorf("tried to half close stream that was already half closed")
+
+type PipeStream struct {
+	inbound  net.Conn
+	outbound net.Conn
+}
+
+func NewPipeStream(inbound net.Conn, outbound net.Conn) *PipeStream {
+	return &PipeStream{
+		inbound:  inbound,
+		outbound: outbound,
+	}
+}
+
+func (s *PipeStream) Close() error {
+	return s.outbound.Close()
+}
+
+func (s *PipeStream) Reset() error {
```

Comment:
Hm. With a normal pipe conn, we'd be able to call `CloseWithError`. Unfortunately, that doesn't look possible here.

We either need to:

1. Re-implement this type.
2. Set some atomic "reset" flag _before_ closing. and check that flag _after_ reading/writing.

Reply:
what exactly do you mean here? the implementation as stands seems reasonable to me

Reply:
Resetting a stream needs to close it with an error. We don't want the other side to think we _successfully_ finished writing, we want them to see that something went wrong. For example, `ioutil.ReadAll(stream)` should fail of the remote side calls `Reset()`.

Reply:
ahhh makes total sense. missed this comment. implemented a simple version of this with buffered channels.

Reply:
updated friday @Stebalien